### ISSUE TITLE
Incorrect premise in readme

### ISSUE THREAD

**Issue by dazld** [patrick-steele-idem/morphdom#60]
Applaud your work, and looks like a great project!

I just want to pick a little hole in your premise for why people use things like react - it's categorically not because the DOM is slow, but more to do with the logic around handling competing updates to a nested tree structure. 

Expanding on this example - a set of nested views. A call to render higher up the tree should (probably) not cause the child views to re-render. It is of course possible to compute this logic, store references, and replace them or not depending on this, but the code to do so is often fragile and has its own series of abstractions. I use react/reactish library x because I don't have to worry about any of this. At no point have I ever considered the DOM to be slow - in fact, I accept the compromise that react is in general _slower_ than direct DOM manipulation because of the lighter mental overhead and stronger abstractions when building complex stuff. 

Does this make sense..?

**Comment by yoshuawuyts** [comment]
Do you have any quotes where that's the case, because it's not what I'm gathering from the README. From the "what about the virtual DOM" section:
> Both morphdom and virtual DOM based solutions update the real DOM with the minimum number of changes. The only difference is in how the differences are determined. morphdom compares real DOM nodes while virtual-dom and others only compare virtual DOM nodes.

Feels pretty far removed from the "why" of using React.

**Comment by dazld** [comment]
I was looking particularly at the line `The premise for using a virtual DOM is that the DOM is "slow".` - just this. Do my points above make sense in this context? Am more than happy to be corrected if I've got something wrong.

**Comment by yoshuawuyts** [comment]
Oh yeah, alright

**Comment by patrick-steele-idem** [comment]
Hey @dazld, I see where you are coming from, but optimizing what part of the tree is rerendered and updated is not limited to a virtual DOM-based solution. You could just as easily be smart about rerendering and diffing/patching only the portions of the real DOM tree that are impacted by state changes.

I consider `morphdom` more of a lower level library for diffing/patching a DOM subtree, but you would still want to use a library on top of `morphdom` to control what gets rerendered and which nodes get diffed/patched. `morphdom` gives you the hooks to skip diffing a DOM subtree or to control how individual nodes get diffed/patched.

I originally created `morphdom` for use with [Marko Widgets](http://markojs.com/docs/marko-widgets/). When using using Marko Widgets, only when a UI component's state changes is the UI component's new DOM rendered. Therefore, only a subset of the DOM is actually being diffed/patched. Portions of the DOM tree that didn't change are just skipped completely. Marko Widget's offers many of the important benefits of React with _a lot less code_ and no virtual DOM and performance has been great for both desktop and mobile web apps.

Assuming you are targeting the real web browser DOM (ignoring React Native...), I believe you can accomplish the same end result using a solution based only on the real DOM instead of introducing a virtual DOM. You would only want to use a virtual DOM-based solution if using the real DOM is "slow". Hence:

> The premise for using a virtual DOM is that the DOM is "slow".

Also, at the time React was introduced there was a lot of talk about the real DOM being too slow to achieve the benefits of React (hence the need for introducing the concept of a virtual DOM). The "DOM is slow" talk has seemed to have died down.

I originally created `morphdom` to compare the performance of the following:

- Diffing/patching a real DOM tree with another real DOM tree
- Diffing/patching a virtual DOM tree with another virtual DOM tree
- Diffing/patching a virtual DOM tree with a real DOM tree (I never got around to this)

I was really just wondering if it was worth considering rendering a virtual DOM tree for the view libraries that I was working on. After creating `morphdom` and various benchmarks, I concluded that rerendering a real DOM tree and diffing the real DOM diffing was very comparable performance and has a lot of benefits over a virtual DOM-based solution. 

What do you think? Do you think we still need to adjust the description of `morphdom` or that its usefulness is still questionable?

**Comment by dazld** [comment]
Great reply, thanks for taking the time @patrick-steele-idem

I think probably just expanding a bit might make it clearer. Instead of the blanket statement which is there right now, perhaps incorporate some of the detail from earlier in the readme, and your reply now. Just for kicks, something like the following..?

```
One of the premises for using a virtual DOM is that performing operations on a real DOM can have unintended side effects, such as reflows and repaints. This (amongst other issues) can end up in a poor user experience. Nothing will prevent a developer from programmatically creating those problems, so the idea that operating on a virtual DOM will magically improve your app performance while performing these operations (which ultimately have to reach out to the real DOM at some point) is a fallacy. As such, real DOM nodes are just as effective as virtual DOM nodes for programmatically modelling an interface structure - unsurprising, considering that is exactly what they are designed to do. 

While there is slightly more overhead in creating actual DOM nodes instead of lightweight virtual DOM nodes, we are not seeing any noticeable slowness in our benchmarks. In addition, as web browsers get faster the DOM data structure will also likely continue to get faster so there benefits to avoiding the abstraction layer.
´´`

Any use?

**Comment by patrick-steele-idem** [comment]
Thanks @dazld. I'm not opposed to updating the docs for more clarity. I'll give it some thought and update soon. Your paragraphs are a good start. Thanks again.

**Comment by AlexGalays** [comment]
To be honest, I think Facebook overdid it marketing wise with all the bullshit speed arguments (react is one of the slowest implementation nowadays).
To me, It seems the most compelling case for virtual nodes is that your "templating" is just JS, you can manipulate the tree in any way you want with an actual programming language, rather than using strings and conventions. This is even more compelling if you use typescript or flow, as you can type a virtual node and component tree, but can't type strings with a lot of logic in them.

**Comment by sandro-pasquali** [comment]
Just want to be clear on this. So I started using marko widgets (and morphdom) because I was sold on the argument that dom<>dom morphing was just as fast as vdom solutions (I may have misunderstood). Am I right in believing that the current view of the team is that using vdom to make dom changes is much faster/better? Just want clarity. I would be a little sad if it turns out this "other" way of doing things turns out to have been a mistake, as vdom is just the better way to think about these things.

**Comment by patrick-steele-idem** [comment]
@sandro-pasquali `morphdom` is plenty fast when diffing the real DOM with the real DOM and at eBay we never saw any performance problems in our webapps and all updates to the DOM were sub 2ms. With that said, in [often contrived] benchmarks `morphdom` didn't fair as well compared to other VDOM solutions so we updated `morphdom` to be compatible with VDOM (see: [morphdom/docs/virtual-dom.md](https://github.com/patrick-steele-idem/morphdom/blob/master/docs/virtual-dom.md)). The [next version of Marko](https://github.com/marko-js/marko/blob/master/ROADMAP.md#virtual-dom-support-366) will be using `morphdom` with VDOM rendering. marko v4 will be a lot faster because we are able to apply additional optimizations when rendering to a VDOM that were not possible when rendering to HTML (that needed to be parsed to a DOM).

__tldr; the DOM is plenty fast, but a VDOM allows additional optimizations. `morphdom` works with both the VDOM and the DOM.__

**Comment by snewcomer** [comment]
These days the IO slows down real dom manipulations on large diffs, so on large diffs the virtual dom is somewhat faster.  For small diffs, it doesn't seem to make a difference with the updated benchmarks.

Check them out and lmk if anybody has any questions!
